ruby/basics/hashes.md

```diff
@@ -1,51 +1,185 @@
 
-what the student is expected to know or be able to do by the end of this lesson
+Basically like arrays, right?
 
-* Learning outcome 1
-* Learning outcome 2
-* Learning outcome 3
+## Learning Outcomes
 
-## Creating a Hash
-* the literal notation with curly braces
-* creating a new hash object with Hash.new
-* code snippets
+By the end of this reading you'll be able to:
 
-## Adding Data
-* How to add a key value pair to the hash
-* code snippets
+* Tell what the difference is between hashes and arrays
+* Know how to create a new hash
+* Know how to populate that hash
+* Know the difference between keys and values
+* Change data within a hash
+* Delete data from a hash
+* Iterate through a hash
+* Merge two hashes together
 
-## Accessing Data
-* How to access data in a hash
-* code snippets
 
-## Changing Data
-* how to change a value in a hash
-* code snippets
+Arrays are maybe best described as bins on a shelf with numbers on them, and each containing something WONDERFUL. E.g. *Bin 0* contains a string that says "zoob", *Bin 1* contains the integer 9, and *Bin 2* contains, weirdly, another bin labeled '*Bin 0*' (but rats, it's empty). This maps out to an array like:
 
-## Removing Data
-* setting the value to nil
-* using the delete method
+```ruby
+array = [ "zoob", 9, [] ]
+array[0]
+  => "zoob"
+array[1]
+  => 9
+array[2]
+  => []
+```
 
-## Merging Hashes
-* how to merge two hashes, explain how it works
-* code snippets
+Hashes take the analogy a step further by allowing you to write the labels on the outside of the bins, which lines up more with how people act in the real world, which is more or less the whole point of Object-Oriented Programming in the first place.
 
-## Useful Methods
-* keys
-* values
+So, back to the shelf. Here's another set of bins, labeled, not 0, 1, and 2, but "random_word", "Dorothy's math test score", "Guess what, literally another bin". Since the labels here aren't numbers in ascending order starting at 0, we can be pretty sure that what we've got here is *a hash*.
 
-## Hash best practices
-* use symbols for keys where possible
+```ruby
+hash = { "random_word" => "ahoy", "Dorothy's math test score" => 94, "Guess what, literally another bin" => {} }
+```
+
+There's a little bit to parse here, so let's take a breather and puzzle these bins out.
+
+One of the first things you might notice is that hashes use *curly braces* `{}` instead of *brackets* `[]`. That's an important one. It helps us distinguish hashes from arrays (useful, since they share a lot of functionality).
+
+The other thing you might notice is that hashes are made up of two parts: *keys* and *values*. A *key* is more or less analogous to an array's *index*, in that it's the label on the outside of the bin that helps us find what we're looking for. The *value* is what gets stored at a particular *key*.
```

Comment:
I wonder if we should move this whole section about creating a hash with the curly bracket notation under the "Creating Hashes" subheading?

This is an excellent description 😄